### PR TITLE
Ensure sub-query use for Job instances.

### PR DESCRIPTION
**Pull request by corneil**
Fix find by name query ordering.
 Fix find by name not found query.
 Improve test case readability.
 #5484

### CODE REVIEW

**Review by onobc — CHANGES_REQUESTED** [review]
Thanks for the PR @corneil . I have a few comments/change requests to take a look at please. 

Also, is there a test that exhibits the bug this is fixing? If not, can we make one or is this a performance related issue?

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java [hunk 1]
```diff
@@ -122,36 +119,46 @@ public void testJobInstanceControllerConstructorMissingRepository() {
 	@Test
 	public void testGetInstanceNotFound() throws Exception {
 		mockMvc.perform(get("/jobs/instances/1345345345345").accept(MediaType.APPLICATION_JSON))
-				.andExpect(status().isNotFound());
+			.andDo(print())
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java:122]: They are good for diagnostic, but once we have diagnosed the issue, we should remove them.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java [hunk 2]
```diff
@@ -122,36 +119,46 @@ public void testJobInstanceControllerConstructorMissingRepository() {
...
+			.andExpect(status().isNotFound());
 	}
 
 	@Test
 	public void testGetInstance() throws Exception {
-		mockMvc.perform(get("/jobs/instances/1").accept(MediaType.APPLICATION_JSON)).andExpect(status().isOk())
-				.andExpect(content().json("{jobInstanceId: " + 1 + "}"));
+		mockMvc.perform(get("/jobs/instances/1").accept(MediaType.APPLICATION_JSON))
+			.andDo(print())
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java:129]: We can remove the print here and below.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskJobService.java [hunk 3]
```diff
@@ -186,7 +186,7 @@ public TaskJobExecution getJobExecution(long id, String schemaTarget) throws NoS
 	}
 
 	@Override
-	public Page<JobInstanceExecutions> listTaskJobInstancesForJobName(Pageable pageable, String jobName) {
+	public Page<JobInstanceExecutions> listTaskJobInstancesForJobName(Pageable pageable, String jobName) throws NoSuchJobException {
```
[the remark this answers is not in the document]
Comment (onobc) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskJobService.java:189]: Understood. Thanks for fixing it.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java [hunk 4]
```diff
@@ -52,22 +49,21 @@
 import org.springframework.test.web.servlet.setup.MockMvcBuilders;
 import org.springframework.web.context.WebApplicationContext;
 
-import static org.hamcrest.Matchers.hasSize;
-import static org.hamcrest.Matchers.is;
+import java.util.ArrayList;
+import java.util.Date;
+
+import static org.hamcrest.Matchers.*;
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java:55]: Missed this wildcard.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java [hunk 5]
```diff
@@ -122,36 +119,46 @@ public void testJobInstanceControllerConstructorMissingRepository() {
...
-				.andExpect(status().isNotFound());
+			.andDo(print())
+			.andExpect(status().isNotFound());
 	}
 
 	@Test
 	public void testGetInstance() throws Exception {
-		mockMvc.perform(get("/jobs/instances/1").accept(MediaType.APPLICATION_JSON)).andExpect(status().isOk())
-				.andExpect(content().json("{jobInstanceId: " + 1 + "}"));
+		mockMvc.perform(get("/jobs/instances/1").accept(MediaType.APPLICATION_JSON))
```
[the remark this answers is not in the document]
Comment (onobc) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java:126]: Discussed offline.

**Review by onobc — APPROVED** [review]
@corneil LGTM - thanks for the updates. There are still a couple of wildcard imports. Once these are adjusted consider this PR approved. I am approving now but please update the imports.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java [hunk 6]
```diff
@@ -238,14 +231,17 @@ public JdbcAggregateJobQueryDao(DataSource dataSource, SchemaService schemaServi
...
 	}
 
 	@Override
-	public Page<JobInstanceExecutions> listJobInstances(String jobName, Pageable pageable) {
+	public Page<JobInstanceExecutions> listJobInstances(String jobName, Pageable pageable) throws NoSuchJobException {
 		int total = countJobExecutions(jobName);
+		if(total == 0) {
+			throw new NoSuchJobException("No Job with that name either current or historic: [" + jobName + "]");
+		}
 		List<JobInstanceExecutions> taskJobInstancesForJobName = total > 0
```
[the remark this answers is not in the document]
Comment (onobc) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java]: @corneil please answer this one.

[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java]: The ternary operator isn't needed and refactored.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java (hunk 2 above)
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/JobInstanceControllerTests.java]: done

**Pull request merged**